src/frontend/parse.ts

```diff
@@ -320,11 +320,39 @@ export function refSource(node: Node, ref?: Token | Node): Node {
 
 export default function parse(code: string) {
   const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
-  parser.feed(code);
+  try {
+    parser.feed(code);
+  } catch (e) {
+    if (e instanceof Error && "token" in e) {
+      const token: Token = (e as any).token;
+      // https://stackoverflow.com/a/72016226/14611638
+      const expected = [
+        ...new Set(
+          ((e as any).message.match(/(?<=A ).*(?= based on:)/g) ?? []).map(
```

Comment:
Really silly that Nearley doesn't have a better way for this, but I guess it isn't designed for good error messages (just good parsing of valid text).